libfwupdplugin/fu-usb-device.c

```diff
@@ -267,13 +269,352 @@ fu_usb_device_open(FuDevice *device, GError **error)
+
+	if (!g_usb_device_control_transfer(priv->usb_device,
+					   G_USB_DEVICE_DIRECTION_DEVICE_TO_HOST,
+					   G_USB_DEVICE_REQUEST_TYPE_VENDOR,
+					   G_USB_DEVICE_RECIPIENT_DEVICE,
+					   dsinfo->vendor_code,	       /* bRequest */
+					   0x0,			       /* wValue */
+					   USB_OS_20_DESCRIPTOR_INDEX, /* wIndex */
+					   buf,
+					   dsinfo->total_length,
+					   &actual_length,
+					   500,
+					   NULL, /* cancellable */
+					   error)) {
+		g_prefix_error(error, "requested vendor code 0x%02x: ", dsinfo->vendor_code);
+		return NULL;
+	}
+	if (dsinfo->total_length != actual_length) {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_INVALID_DATA,
+			    "expected 0x%x bytes from vendor code 0x%02x, but got 0x%x",
+			    dsinfo->total_length,
+			    dsinfo->vendor_code,
+			    (guint)actual_length);
+		return NULL;
+	}
+
+	/* debug */
+	if (g_getenv("FWUPD_VERBOSE") != NULL)
+		fu_dump_raw(G_LOG_DOMAIN, "PlatformCapabilityOs20", buf, actual_length);
+
+	/* success */
+	return g_bytes_new_take(g_steal_pointer(&buf), actual_length);
+}
+
+static gboolean
+fu_usb_device_parse_platform_capability_data(FuUsbDevice *self,
+					     GBytes *blob,
+					     const gchar *guid,
+					     GError **error)
+{
+	gsize bufsz = 0;
+	const guint8 *buf = g_bytes_get_data(blob, &bufsz);
+	g_autoptr(GPtrArray) dsinfos = g_ptr_array_new_with_free_func(g_free);
+
+	/* debug */
+	if (g_getenv("FWUPD_VERBOSE") != NULL)
+		fu_dump_bytes(G_LOG_DOMAIN, "PlatformCapabilityData", blob);
+	for (gsize off = 0; off < bufsz; off += sizeof(FuUsbDeviceBosPlatformCapabilityDsinfo)) {
+		FuUsbDeviceBosPlatformCapabilityDsinfo *dsinfo =
+		    g_new0(FuUsbDeviceBosPlatformCapabilityDsinfo, 1);
+		g_ptr_array_add(dsinfos, dsinfo);
+		if (!fu_memread_uint32_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, platform_ver),
+			&dsinfo->platform_ver,
+			G_LITTLE_ENDIAN,
+			error))
+			return FALSE;
+		if (!fu_memread_uint16_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, total_length),
+			&dsinfo->total_length,
+			G_LITTLE_ENDIAN,
+			error))
+			return FALSE;
+		if (!fu_memread_uint8_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, vendor_code),
+			&dsinfo->vendor_code,
+			error))
+			return FALSE;
+		if (!fu_memread_uint8_safe(
+			buf,
+			bufsz,
+			off + G_STRUCT_OFFSET(FuUsbDeviceBosPlatformCapabilityDsinfo, alt_code),
+			&dsinfo->alt_code,
+			error))
+			return FALSE;
+		g_debug("PlatformVersion=0x%08x, TotalLength=0x%04x, VendorCode=0x%02x, "
+			"AltCode=0x%02x",
+			dsinfo->platform_ver,
+			dsinfo->total_length,
+			dsinfo->vendor_code,
+			dsinfo->alt_code);
+	}
+
+	/* sort by platform_ver, highest first */
+	g_ptr_array_sort(dsinfos, fu_usb_device_sort_by_dsinfo_platform_version_cb);
+
+	/* use Windows 8.1 as a baseline */
+	if (g_strcmp0(guid, USB_DEVICE_PLATFORM_CAPABILITY_UUID_MICROSOFT) == 0) {
```

Comment:
I don't think you want to mix and match with the Microsoft one.  If this is going to be a program tied specifically to WWCB/LVFS certifications then it needs to be all in on a fwupd GUID.

Reply:
Right, this was added for the future; there are a few fields in the MS descriptor that mean we don't have to assign the composite devices (e.g. docks) in such a manual way. i.e. it's useful, but not for this PR.

Reply:
Shouldn't it be pulled out then?

Reply:
I split it out into a new file.